File: client/src/features/grid.ts
import { Resource } from "../types/types";
import { Location, Tile as TileType } from "../types/types";

// randomizeResource()
// Generates a random Resource object with specific constraints.
export const randomizeResource = (): Resource => {
  // Possible resource types
  const types = ["empty", "stone", "wood"];

  // Randomly select a type from the available types
  const randomType = types[Math.floor(Math.random() * types.length)];

  // Generates a random number within the specified range.
  const getRandomNumberInRange = (min: number, max: number): number => {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  };

  // Generate random values for quantity and quality within the specified range
  const quantity = getRandomNumberInRange(1, 5);
  const quality = getRandomNumberInRange(1, 5);

  // Return the randomly generated Resource object
  return {
    type: randomType,
    quantity,
    quality,
  };
};

// generateGrid(number)
// Generates a grid of tiles based on the specified size.
export const generateGrid = (size: number): TileType[] => {
  // Array to store the generated tiles
  const tiles: TileType[] = [];

  // Loop through rows
  for (let x = 0; x < size; x++) {
    // Loop through columns
    for (let y = 0; y < size; y++) {
      // Create a location object with x and y coordinates
      const location: Location = { x, y };

      // Create a Tile object with a random resource, unoccupied by default
      const tile: TileType = {
        location,
        resource: randomizeResource(),
        occupied: false,
      };

      // Add the tile to the array of tiles
      tiles.push(tile);
    }
  }

  // Return the array of generated tiles
  return tiles;
};

File: server/game/grid.go
package game

import (
	"math/rand"
	"server/types"
	"time"
)

// CreateGrid generates a grid of tiles based on the given size.
func CreateGrid(size int) [][]types.Tile {
	rand.Seed(time.Now().UnixNano())

	// Adjust the size to account for the negative coordinates
	size += 5

	// Create a 2D slice representing the grid
	grid := make([][]types.Tile, size)
	for i := range grid {
		grid[i] = make([]types.Tile, size)
	}

	// Populate the grid with tiles
	for x := -5; x < size-5; x++ {
		for y := -5; y < size-5; y++ {
			// If the coordinates are negative, assign a "None" resource
			if x < 0 || y < 0 {
				grid[x+5][y+5] = types.Tile{
					Location: types.Location{X: x, Y: y},
					Resource: types.Resource{Type: "None", Quantity: 0, Quality: 0},
				}
			} else {
				grid[x+5][y+5] = types.Tile{
					Location: types.Location{X: x, Y: y},
					Resource: GenerateRandomResource(),
				}
			}
		}
	}

	return grid
}

// GenerateRandomResource generates a random resource for a tile.
func GenerateRandomResource() types.Resource {
	resources := []types.Resource{
		{Type: "Wood", Quantity: rand.Intn(10) + 1, Quality: rand.Intn(5) + 1},
		{Type: "Stone", Quantity: rand.Intn(5) + 1, Quality: rand.Intn(5) + 1},
		{Type: "Ore", Quantity: rand.Intn(3) + 1, Quality: rand.Intn(5) + 1},
		// Add more resource types as needed
	}

	return resources[rand.Intn(len(resources))]
}
